src/sources/statsd/mod.rs

```diff
@@ -50,75 +105,102 @@ impl SourceConfig for StatsdConfig {
-
-            info!(
-                message = "Listening.",
-                addr = %addr,
-                r#type = "udp"
-            );
-
-            let _ = UdpFramed::new(socket, BytesCodec::new())
-                .take_until(shutdown.compat())
-                .filter_map(|frame| async move {
-                    match frame {
-                        Ok((bytes, _sock)) => {
-                            let packet = String::from_utf8_lossy(bytes.as_ref());
-                            let metrics = packet
-                                .lines()
-                                .filter_map(|line| match parse(line) {
-                                    Ok(metric) => {
-                                        emit!(StatsdEventReceived {
-                                            byte_size: line.len()
-                                        });
-                                        Some(Ok(Event::Metric(metric)))
-                                    }
-                                    Err(error) => {
-                                        emit!(StatsdInvalidRecord { error, text: line });
-                                        None
-                                    }
-                                })
-                                .collect::<Vec<_>>();
-                            Some(stream::iter(metrics))
-                        }
-                        Err(error) => {
-                            emit!(StatsdSocketError::read(error));
-                            None
-                        }
-                    }
-                })
-                .flatten()
-                .forward(out.sink_compat())
-                .await;
-
-            info!("Finished sending");
-            Ok(())
+    let socket = UdpSocket::bind(&config.address)
+        .map_err(|error| emit!(StatsdSocketError::bind(error)))
+        .await?;
+
+    info!(
+        message = "Listening.",
+        addr = %config.address,
+        r#type = "udp"
+    );
+
+    let _ = UdpFramed::new(socket, BytesCodec::new())
+        .take_until(shutdown.compat())
+        .filter_map(|frame| {
+            future::ready(match frame {
+                Ok((bytes, _sock)) => {
+                    let packet = String::from_utf8_lossy(bytes.as_ref());
+                    let metrics = packet
+                        .lines()
+                        .filter_map(|line| match parse(line) {
+                            Ok(metric) => {
+                                emit!(StatsdEventReceived {
+                                    byte_size: line.len()
+                                });
+                                Some(Ok(Event::Metric(metric)))
+                            }
+                            Err(error) => {
+                                emit!(StatsdInvalidRecord { error, text: line });
+                                None
+                            }
+                        })
+                        .collect::<Vec<_>>();
```

Comment:
I see that this came from the current code, but this `collect` looks excess?

Reply:
Agreed. `stream::iter` below takes an iterator, which is everything up to the `.collect` above.

Reply:
Not possible because of `returns a value referencing data owned by the current function`.

I'm using async/await which does not need `collect`.

Reply:
Indeed, it's because `Cow` from `from_utf8_lossy` can borrow `bytes` which stays in the function. Thanks that fixed :+1: